### PR TITLE
`@Assign` annotation checker

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR introduces a new module called `model-verifier`.

Currently, the `@Assign` annotation is the only one processed by the module.

The processing includes:
 - check for at least one argument presence for the annotated methods;
 - check the argument to be unique in the whole project classpath (i.e. one command handler for one command type).

To enable the check, the framework user should add the following to his `build.gradle` file.

```groovy

dependencies {
    // ...
    compile group: "io.spine.tools", name: "annotation-checker", version: spineVersion
}

compileJava {
    // ...
    options.compilerArgs += ["-processor", "io.spine.tools.model.AssignChecker"]
}

```

In future, it's possible to add a Gradle plugin which would include the required compiler arguments automatically.

### CODE REVIEW

**Comment by armiol** [comment]
@dmdashenkov the module should probably be named with extensibility in mind. Its goal is to verify the application model (and probably prepare some basis for code generation, performed by the `model-compiler`. I would name the module (and the artifact) `model-verifier` (i.e. verifies the integrity of the data model).

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,9 @@
 # Gradle interim configs
 .gradle/*
 
+# Test modules Gradle interim configs
+/*/*/*/.gradle/
```
Comment (armiol) [inline .gitignore]: I don't get it. Let's discuss.

File: annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * Methods designed for inheritance
+     ***********************************/
+
+    /**
+     * Retrieves the supported annotation type of this processor.
+     *
+     * <p>Basic implementation example is:
+     * <pre>
+     *     {@code
+     *     {@literal @Override}
```
Comment (armiol) [inline annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java]: Isn't it ` {@literal @}Override`?

File: annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    @SuppressWarnings("FieldAccessedSynchronizedAndUnsynchronized")
+        // Initialized in the synchronized `init` method.
+    private Messager messager;
+
+    @SuppressWarnings("FieldAccessedSynchronizedAndUnsynchronized")
+        // Initialized in the synchronized `init` method.
+    private Types typeModel;
+
+    /*
+     * Methods designed for inheritance
```
Comment (armiol) [inline annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java]: Please remove that. It's obvious.

File: annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+     *
+     * <p>Does nothing by default. Override this method to change the processing flow.
+     */
+    @SuppressWarnings("NoopMethodInAbstractClass")
+    protected void onRoundFinished() {
+        // NoOp
+    }
+
+    /*
+     * Overridden methods
```
Comment (armiol) [inline annotation-checker/src/main/java/io/spine/tools/model/SpineAnnotationProcessor.java]: Why do we even do these sections? We aren't paid for symbols.

Please remove.

File: annotation-checker/testmodules/duplicate/src/main/java/io/spine/tools/test/MyFirstAggregate.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import io.spine.validate.AnyVBuilder;
+
+import java.util.List;
+
+import static java.util.Collections.singletonList;
+
+/**
+ * @author Dmytro Dashenkov
+ */
+public class MyFirstAggregate extends Aggregate<String, Any, AnyVBuilder> {
```
Comment (armiol) [inline annotation-checker/testmodules/duplicate/src/main/java/io/spine/tools/test/MyFirstAggregate.java]: The files like this one should belong to `test/resources`.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

Concerning the JAR dependencies analysis, this might be a subject of further thinking.

**Comment by armiol** [comment]
@dmdashenkov >> Concerning the JAR dependencies analysis, this might be a subject of further thinking.

Please do so by creating a GitHub issue.

**Comment by armiol** [comment]
@dmdashenkov please make sure you haven't missed any `package-info.java` etc in the newly created module. And feel free to merge.

**Pull request merged**